### PR TITLE
gui: Do not create payment server if -disablewallet option provided

### PR DESCRIPTION
**Pull request by hebasto**
This PR makes early check for the `-disablewallet` option.

If `-disablewallet=1`, objects `PaymentServer` and `WalletController` are  nor created.

### CODE REVIEW

File: src/qt/bitcoin.cpp [hunk 1]
```diff
@@ -211,11 +212,15 @@ BitcoinApplication::~BitcoinApplication()
     delete window;
     window = nullptr;
 #ifdef ENABLE_WALLET
-    delete paymentServer;
-    paymentServer = nullptr;
-    delete m_wallet_controller;
-    m_wallet_controller = nullptr;
-#endif
+    if (paymentServer) {
```
Comment (promag) [inline src/qt/bitcoin.cpp]: This is not necessary, it is safe to `delete nullptr`.

File: src/qt/bitcoin.cpp [hunk 2]
```diff
@@ -331,24 +336,21 @@ void BitcoinApplication::initializeResult(bool success)
...
-#endif
-
         clientModel = new ClientModel(m_node, optionsModel);
         window->setClientModel(clientModel);
 #ifdef ENABLE_WALLET
-        window->setWalletController(m_wallet_controller);
+        if (WalletModel::isWalletEnabled()) {
+            m_wallet_controller = new WalletController(m_node, platformStyle, optionsModel, this);
+            window->setWalletController(m_wallet_controller);
+            if (paymentServer) {
```
Comment (promag) [inline src/qt/bitcoin.cpp]: This could be `assert(paymentServer)`?

File: src/qt/bitcoin.cpp (hunk 1 above)
Comment (promag) [reply src/qt/bitcoin.cpp]: Or maybe you want to get rid of these `delete` as these objects are owned by `BitcoinApplication` instance.

**Review by promag — COMMENTED** [review]
Concept ACK, makes sense to not instantiate these objects when `-disablewallet`.

It would be nice to remove all `ENABLE_WALLET` and other similar checks from `src/qt/bitcoin.cpp`.

nit, commit could also have prefix.

File: src/qt/bitcoin.cpp (hunk 2 above)
Comment (hebasto) [reply src/qt/bitcoin.cpp]: Failing to create a PaymentServer object is not critical, IMO.

File: src/qt/bitcoin.cpp (hunk 1 above)
Comment (hebasto) [reply src/qt/bitcoin.cpp]: > Or maybe you want to get rid of these `delete` as these objects are owned by `BitcoinApplication` instance.

Exactly!

**Comment by hebasto** [comment]
@promag Thank you for your review. Your comments have been addressed.

> It would be nice to remove all `ENABLE_WALLET` and other similar checks from `src/qt/bitcoin.cpp`.

Could you elaborate your suggestion?

**Comment by fanquake** [comment]
Concept ACK

**Comment by jonasschnelli** [comment]
utACK 4057b7acb7125739537078d026ad96bb21708e3c
Removes three #ifdefs, probably very tiny ressource reduction when not running with the wallet.

File: src/qt/bitcoin.cpp (hunk 2 above)
Comment (laanwj) [reply src/qt/bitcoin.cpp]: Also `new` will never return a `nullptr` anyway.

**Comment by laanwj** [comment]
ACK 4057b7acb7125739537078d026ad96bb21708e3c

**Pull request merged**